File: SRC/dlarfgp.f
      SUBROUTINE DLARFGP( N, ALPHA, X, INCX, TAU )

*  -- LAPACK auxiliary routine --
*  -- LAPACK is a software package provided by Univ. of Tennessee,    --
*  -- Univ. of California Berkeley, Univ. of Colorado Denver and NAG Ltd..--

      // .. Scalar Arguments ..
      int                INCX, N;
      double             ALPHA, TAU;
      // ..
      // .. Array Arguments ..
      double             X( * );
      // ..

*  =====================================================================

      // .. Parameters ..
      double             TWO, ONE, ZERO;
      const              TWO = 2.0D+0, ONE = 1.0D+0, ZERO = 0.0D+0 ;
      // ..
      // .. Local Scalars ..
      int                J, KNT;
      double             BETA, BIGNUM, EPS, SAVEALPHA, SMLNUM, XNORM;
      // ..
      // .. External Functions ..
      double             DLAMCH, DLAPY2, DNRM2;
      // EXTERNAL DLAMCH, DLAPY2, DNRM2
      // ..
      // .. Intrinsic Functions ..
      // INTRINSIC ABS, SIGN
      // ..
      // .. External Subroutines ..
      // EXTERNAL DSCAL
      // ..
      // .. Executable Statements ..

      IF( N.LE.0 ) THEN
         TAU = ZERO
         RETURN
      END IF

      EPS = DLAMCH( 'Precision' )
      XNORM = DNRM2( N-1, X, INCX )

      IF( XNORM.LE.EPS*ABS(ALPHA) ) THEN

         // H  =  [+/-1, 0; I], sign chosen so ALPHA >= 0.

         IF( ALPHA.GE.ZERO ) THEN
            // When TAU.eq.ZERO, the vector is special-cased to be
            // all zeros in the application routines.  We do not need
           t // o clear it.
            TAU = ZERO
         ELSE
            // However, the application routines rely on explicit
            // zero checks when TAU.ne.ZERO, and we must clear X.
            TAU = TWO
            DO J = 1, N-1
               X( 1 + (J-1)*INCX ) = 0
            END DO
            ALPHA = -ALPHA
         END IF
      ELSE

         // general case

         BETA = SIGN( DLAPY2( ALPHA, XNORM ), ALPHA )
         SMLNUM = DLAMCH( 'S' ) / DLAMCH( 'E' )
         KNT = 0
         IF( ABS( BETA ).LT.SMLNUM ) THEN

            // XNORM, BETA may be inaccurate; scale X and recompute them

            BIGNUM = ONE / SMLNUM
   10       CONTINUE
            KNT = KNT + 1
            CALL DSCAL( N-1, BIGNUM, X, INCX )
            BETA = BETA*BIGNUM
            ALPHA = ALPHA*BIGNUM
            IF( (ABS( BETA ).LT.SMLNUM) .AND. (KNT .LT. 20) ) GO TO 10

            // New BETA is at most 1, at least SMLNUM

            XNORM = DNRM2( N-1, X, INCX )
            BETA = SIGN( DLAPY2( ALPHA, XNORM ), ALPHA )
         END IF
         SAVEALPHA = ALPHA
         ALPHA = ALPHA + BETA
         IF( BETA.LT.ZERO ) THEN
            BETA = -BETA
            TAU = -ALPHA / BETA
         ELSE
            ALPHA = XNORM * (XNORM/ALPHA)
            TAU = ALPHA / BETA
            ALPHA = -ALPHA
         END IF

         IF ( ABS(TAU).LE.SMLNUM ) THEN

            // In the case where the computed TAU ends up being a denormalized number,
            // it loses relative accuracy. This is a BIG problem. Solution: flush TAU
           t // o ZERO. This explains the next IF statement.

            // (Bug report provided by Pat Quillen from MathWorks on Jul 29, 2009.)
            // (Thanks Pat. Thanks MathWorks.)

            IF( SAVEALPHA.GE.ZERO ) THEN
               TAU = ZERO
            ELSE
               TAU = TWO
               DO J = 1, N-1
                  X( 1 + (J-1)*INCX ) = 0
               END DO
               BETA = -SAVEALPHA
            END IF

         ELSE

            // This is the general case.

            CALL DSCAL( N-1, ONE / ALPHA, X, INCX )

         END IF

         // If BETA is subnormal, it may lose relative accuracy

         DO 20 J = 1, KNT
            BETA = BETA*SMLNUM
 20      CONTINUE
         ALPHA = BETA
      END IF

      RETURN

      // End of DLARFGP

      }

File: lib/src/dlarfgp.dart
// Copyright (c) 2024 Guilherme Lepsch. All rights reserved. Use of this
// source code is governed by a BSD-style license that can be found in the
// [LICENSE file](https://github.com/lepsch/dart_lapack/blob/main/LICENSE).

import 'package:dart_lapack/src/blas/dnrm2.dart';
import 'package:dart_lapack/src/blas/dscal.dart';
import 'package:dart_lapack/src/box.dart';
import 'package:dart_lapack/src/dlapy2.dart';
import 'package:dart_lapack/src/install/dlamch.dart';
import 'package:dart_lapack/src/intrinsics/sign.dart';
import 'package:dart_lapack/src/matrix.dart';

void dlarfgp(
  final int N,
  final Box<double> ALPHA,
  final Array<double> X_,
  final int INCX,
  final Box<double> TAU,
) {
  final X = X_.having();
  const TWO = 2.0, ONE = 1.0, ZERO = 0.0;
  int J, KNT = 0;
  double BETA = 0, BIGNUM = 0, EPS, SAVEALPHA, SMLNUM = 0, XNORM;

  if (N <= 0) {
    TAU.value = ZERO;
    return;
  }

  EPS = dlamch('Precision');
  XNORM = dnrm2(N - 1, X, INCX);

  if (XNORM <= EPS * ALPHA.value.abs()) {
    // H  =  [+/-1, 0; I], sign chosen so ALPHA >= 0.
    if (ALPHA.value >= ZERO) {
      // When TAU == ZERO, the vector is special-cased to be
      // all zeros in the application routines.  We do not need
      // to clear it.
      TAU.value = ZERO;
    } else {
      // However, the application routines rely on explicit
      // zero checks when TAU != ZERO, and we must clear X.
      TAU.value = TWO;
      for (J = 1; J <= N - 1; J++) {
        X[1 + (J - 1) * INCX] = 0;
      }
      ALPHA.value = -ALPHA.value;
    }
  } else {
    // general case
    BETA = sign(dlapy2(ALPHA.value, XNORM), ALPHA.value);
    SMLNUM = dlamch('S') / dlamch('E');
    KNT = 0;
    if (BETA.abs() < SMLNUM) {
      // XNORM, BETA may be inaccurate; scale X and recompute them
      BIGNUM = ONE / SMLNUM;
      do {
        KNT++;
        dscal(N - 1, BIGNUM, X, INCX);
        BETA *= BIGNUM;
        ALPHA.value *= BIGNUM;
      } while ((BETA.abs() < SMLNUM) && (KNT < 20));

      // New BETA is at most 1, at least SMLNUM
      XNORM = dnrm2(N - 1, X, INCX);
      BETA = sign(dlapy2(ALPHA.value, XNORM), ALPHA.value);
    }
    SAVEALPHA = ALPHA.value;
    ALPHA.value += BETA;
    if (BETA < ZERO) {
      BETA = -BETA;
      TAU.value = -ALPHA.value / BETA;
    } else {
      ALPHA.value = XNORM * (XNORM / ALPHA.value);
      TAU.value = ALPHA.value / BETA;
      ALPHA.value = -ALPHA.value;
    }

    if (TAU.value.abs() <= SMLNUM) {
      // In the case where the computed TAU ends up being a denormalized number,
      // it loses relative accuracy. This is a BIG problem. Solution: flush TAU
      // to ZERO. This explains the next IF statement.
      if (SAVEALPHA >= ZERO) {
        TAU.value = ZERO;
      } else {
        TAU.value = TWO;
        for (J = 1; J <= N - 1; J++) {
          X[1 + (J - 1) * INCX] = 0;
        }
        BETA = -SAVEALPHA;
      }
    } else {
      // This is the general case.
      dscal(N - 1, ONE / ALPHA.value, X, INCX);
    }

    // If BETA is subnormal, it may lose relative accuracy
    for (J = 1; J <= KNT; J++) {
      BETA *= SMLNUM;
    }
    ALPHA.value = BETA;
  }
}
